File: app/src/main/java/com/damianmichalak/shopping_list/view/ProductsListManager.java
package com.damianmichalak.shopping_list.view;


import android.content.res.Resources;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.damianmichalak.shopping_list.R;
import com.damianmichalak.shopping_list.helper.DateHelper;
import com.damianmichalak.shopping_list.model.api_models.Product;
import com.damianmichalak.shopping_list.presenter.ProductsListPresenter;
import com.jacekmarchwicki.universaladapter.BaseAdapterItem;
import com.jacekmarchwicki.universaladapter.ViewHolderManager;
import com.jakewharton.rxbinding.view.RxView;

import javax.annotation.Nonnull;
import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import rx.subscriptions.SerialSubscription;
import rx.subscriptions.Subscriptions;

public class ProductsListManager implements ViewHolderManager {

    @Nonnull
    private final Resources resources;
    @Nonnull
    private final DateHelper dateHelper;

    @Inject
    public ProductsListManager(@Nonnull Resources resources, @Nonnull DateHelper dateHelper) {
        this.resources = resources;
        this.dateHelper = dateHelper;
    }

    @Override
    public boolean matches(@Nonnull BaseAdapterItem baseAdapterItem) {
        return baseAdapterItem instanceof ProductsListPresenter.ShoppingListItemWithKey;
    }

    @Nonnull
    @Override
    public BaseViewHolder createViewHolder(@Nonnull ViewGroup parent, @Nonnull LayoutInflater inflater) {
        return new ProductViewHolder(inflater.inflate(R.layout.list_item_shopping, parent, false));
    }

    class ProductViewHolder extends BaseViewHolder {

        @BindView(R.id.shopping_item_name)
        TextView itemName;
        @BindView(R.id.shopping_item_date)
        TextView itemDate;
        @BindView(R.id.shopping_item_done)
        View done;

        @Nonnull
        private final SerialSubscription subscription = new SerialSubscription();

        public ProductViewHolder(@Nonnull View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }

        @Override
        public void onViewRecycled() {
            super.onViewRecycled();
            subscription.set(Subscriptions.empty());
        }

        @Override
        public void bind(@Nonnull BaseAdapterItem item) {
            final ProductsListPresenter.ShoppingListItemWithKey adapterItem = (ProductsListPresenter.ShoppingListItemWithKey) item;
            final Product product = adapterItem.getProduct();
            itemName.setText(product.getName());
            itemDate.setText(resources.getString(R.string.shopping_list_date_string, dateHelper.getDateForTimestamp(product.getDateAdded())));

            subscription.set(Subscriptions.from(
                    RxView.clicks(done)
                            .subscribe(adapterItem.removeItem())
            ));
        }

    }

}

File: app/src/main/java/com/damianmichalak/shopping_list/view/ProductsListManager.kt
package com.damianmichalak.shopping_list.view


import android.content.res.Resources
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.damianmichalak.shopping_list.R
import com.damianmichalak.shopping_list.helper.DateHelper
import com.damianmichalak.shopping_list.presenter.ProductsListPresenter
import com.jacekmarchwicki.universaladapter.BaseAdapterItem
import com.jacekmarchwicki.universaladapter.ViewHolderManager
import com.jakewharton.rxbinding.view.RxView
import kotlinx.android.synthetic.main.list_item_shopping.view.*
import rx.subscriptions.SerialSubscription
import rx.subscriptions.Subscriptions
import javax.inject.Inject

class ProductsListManager @Inject
constructor(private val resources: Resources, private val dateHelper: DateHelper) : ViewHolderManager {

    override fun matches(baseAdapterItem: BaseAdapterItem): Boolean {
        return baseAdapterItem is ProductsListPresenter.ShoppingListItemWithKey
    }

    override fun createViewHolder(parent: ViewGroup, inflater: LayoutInflater): ViewHolderManager.BaseViewHolder<ProductsListPresenter.ShoppingListItemWithKey> {
        return ProductViewHolder(inflater.inflate(R.layout.list_item_shopping, parent, false))
    }

    internal inner class ProductViewHolder(val view: View) : ViewHolderManager.BaseViewHolder<ProductsListPresenter.ShoppingListItemWithKey>(view) {


        //        private final Observable<Void> clickObservable;
        private val subscription = SerialSubscription()


        override fun onViewRecycled() {
            super.onViewRecycled()
//            subscription.set(Subscriptions.empty());
        }

        override fun bind(item: ProductsListPresenter.ShoppingListItemWithKey) {
            val product = item.product
            view.shopping_item_name.text = product.name
            view.shopping_item_date.text = resources.getString(R.string.shopping_list_date_string, dateHelper.getDateForTimestamp(product.dateAdded))

            subscription.set(Subscriptions.from(
                    RxView.clicks(view.shopping_item_done).subscribe(item.removeItem())
            ))
        }

    }

}
